lib/reservation_tracker.rb

```diff
@@ -0,0 +1,267 @@
+require 'csv'
+
+require_relative 'date_range'
+require_relative 'reservation'
+require_relative 'room'
+require_relative 'block'
+
+NUM_OF_ROOMS = 20
+MAX_BLOCK_NUM = 5
+
+module Hotel
+  class ReservationTracker
+    attr_reader :rooms, :blocks, :reservations
+
+    def initialize(room_file = 'support/rooms.csv', block_file = 'support/blocks.csv', reservation_file = 'support/reservations.csv')
+      @rooms = load_rooms(room_file)
+      @blocks = load_blocks(block_file)
+      @reservations = load_reservations(reservation_file)
+    end
+
+    def load_rooms(filename)
+      rooms = []
+
+      CSV.read(filename, headers: true).each do |line|
+
+        input_data = {}
+        input_data[:room_num] = line[0].to_i
+        input_data[:rate] = line[1].to_i
+
+        room = Room.new(input_data)
+        rooms << room
+      end
+      return rooms
+    end
+
+    def load_blocks(filename)
+      blocks = []
+      block_data = CSV.open(filename, 'r', headers: true, header_converters: :symbol)
+
+      block_data.each do |raw_block|
+        start_date = Date.parse(raw_block[:start_date])
+        end_date = Date.parse(raw_block[:end_date])
+
+        requested_qty = raw_block[:party].to_i
+        block_qty_rooms = @rooms.take(requested_qty)
+        requested_dates = DateRange.new(start_date, end_date)
+
+        parsed_block = {
+          id: raw_block[:id].to_i,
+          party: block_qty_rooms,
+          date_range: requested_dates
+        }
+
+        block = Block.new(parsed_block)
+        blocks << block
+      end
+      return blocks
+    end
+
+    def load_reservations(filename)
+      reservations = []
+      reservation_data = CSV.open(filename, 'r', headers: true, header_converters: :symbol)
+
+      reservation_data.each do |raw_reservation|
+        block_id = find_block_id(raw_reservation[:block_id].to_i)
+        room = find_room(raw_reservation[:room].to_i)
+
+        start_date = Date.parse(raw_reservation[:start_date])
+        end_date = Date.parse(raw_reservation[:end_date])
+
+        range = DateRange.new(start_date, end_date)
+
+        parsed_reservation = {
+          id: raw_reservation[:id].to_i,
+          block_id: block_id,
+          room: room,
+          date_range: range
+        }
+
+        reservation = Reservation.new(parsed_reservation)
+        reservations << reservation
+      end
+      return reservations
+    end
+
+    def find_block_id(id)
+      check_id(id)
+      return @blocks.find { |block| block.id == id }
+    end
```

Comment:
The name of this method is a little misleading: it finds a block _by_ ID, it does not find a block ID. Something like `find_block` or `find_block_by_id` might be better.